src/builder.rs

```diff
@@ -0,0 +1,472 @@
+                sighash,
+            },
+        )
+    }
+
+    /// Applies signatures to this bundle, in order to authorize it.
+    pub fn apply_signatures<R: rand_7::RngCore + rand_7::CryptoRng>(
+        self,
+        mut rng: R,
+        sighash: [u8; 32],
+        signing_keys: &[SpendAuthorizingKey],
+    ) -> Result<Bundle<Authorized>, Error> {
+        signing_keys
+            .iter()
+            .fold(self.prepare(&mut rng, sighash), |partial, ask| {
+                partial.sign(&mut rng, ask)
+            })
+            .finalize()
+    }
+}
+
+impl Bundle<PartiallyAuthorized> {
+    /// Signs this bundle with the given [`SpendAuthorizingKey`].
+    ///
+    /// This will apply signatures for all notes controlled by this spending key.
+    pub fn sign<R: rand_7::RngCore + rand_7::CryptoRng>(
+        self,
+        mut rng: R,
+        ask: &SpendAuthorizingKey,
+    ) -> Self {
+        let expected_ak = ask.into();
+        self.authorize(
+            &mut rng,
+            |rng, partial, (sig, ak)| {
+                (
+                    sig.or_else(|| {
+                        if ak == expected_ak {
+                            Some(ask.sign(rng, &partial.sighash))
+                        } else {
+                            None
+                        }
+                    }),
+                    ak,
+                )
+            },
+            |_, partial| partial,
+        )
+    }
+
+    /// Finalizes this bundle, enabling it to be included in a transaction.
+    ///
+    /// Returns an error if any signatures are missing.
+    pub fn finalize(self) -> Result<Bundle<Authorized>, Error> {
+        self.try_authorize(
+            &mut (),
+            |_, _, (sig, _)| match sig {
+                Some(sig) => Ok(sig),
+                None => Err(Error::MissingSignatures),
+            },
+            |_, partial| {
+                Ok(Authorized::from_parts(
+                    partial.proof,
+                    partial.binding_signature,
+                ))
+            },
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::rngs::OsRng;
+
+    use super::Builder;
+    use crate::{
+        bundle::Flags,
+        circuit::ProvingKey,
+        keys::{FullViewingKey, SpendingKey},
+        tree::Anchor,
+        value::{NoteValue, ValueSum},
+    };
+
+    #[test]
+    fn shielding_bundle() {
+        let pk = ProvingKey::build();
+        let mut rng = OsRng;
+
+        let sk = SpendingKey::random(&mut rng);
+        let fvk = FullViewingKey::from(&sk);
+        let recipient = fvk.default_address();
+
+        let mut builder = Builder::new(Flags::from_parts(true, true), Anchor);
+        builder
+            .add_recipient(None, recipient, NoteValue::from_raw(5000), None)
+            .unwrap();
+        let bundle = dbg!(builder
+            .build(&mut rng, &pk)
+            .unwrap()
+            .prepare(rand_7::rngs::OsRng, [0; 32]))
+        .finalize()
```

Comment:
Should this be calling `sign` before finalization?

Reply:
No, because this test has no real spent notes. Once we have note encryption support (and thus can get the created notes out of a builder), we can extend the tests with a case that exercises note spending.